### PR TITLE
Create FERC-cleaning analysis module

### PR DESCRIPTION
**Pull request by aesharpe**
# FERC is a mess and we all know it...

This PR introduces a ferc-cleaning analysis module called `clean_up_ferc1`. Eventually this module will contain functions to clean up other tables from FERC form one, but right now it only contains cleaning functions for the `small_generators` table (I have some stuff lined up for the steam table).

I've also created a notebook called `clean_ferc1_analysis.ipynb` that walks you though each step of this analysis and allows you to take a closer look at the data outputs from each step. 

I have NOT (yet) incorporated this module into any of the FERC output tables. The only place to explore it right now is in the notebook.

## Module Contents:

The `clean_up_ferc1` module addresses some of the key issues with the small generators table. It is comprised of six steps:

1) **Remove obviously bad rows** (where there is no data and there are NA or ---- names)
2) **Label row type** (total, note, header, or NA for regular rows)
3) **Use headers and clues in plant name to supplement plant type column** (currently kept separate for comparison purposes -- `plant_type` for Zane's manual effort and `plant_type_2` for my programatic scrape).
4) **Extract FERC license number from plant name** (and put them in a column called `ferc_license` not to be confused with Zane's manual `ferc_license_manual` column kept for comparison purposes.
5) **Associate notes rows with their respective value rows** (this also applies to FERC license numbers that appear in the notes. The association is done via the presence of footnotes like (a), (b), etc. and therefore does not catch and associate every note row.
6) **Remove header and note rows from the final df** (with an optional flag to remove totals or not).

### Module Strengths:

The module does a good job of identifying row types (header, note, total, NA aka regular), extracting FERC license numbers, and associating notes with their corresponding value rows.

### Module Weaknesses:

The module does a pretty good job of associating header fuel types with the correct group of records, but there is a chance that some of these associations are incorrect either due to human error with data entry in the form or programatic misinterpretation of the data. This is a problem for those looking for highly accurate plant type information. I don't think that this is a huge problem because we are currently working with RMI to match individual FERC and EIA records which would result in a solid swath of FERC records now having an EIA technology description. We could use this field for comparison, but I'm not sure how to treat it in the short run. The column I've put these cleaned and associated headers in is called `plant_type_2`

The module occasionally mis-reads certain numbers as FERC licenses (I do have a safeguard so it doesn't do this with years).

### Possible Additions: 

Right now, I'm grabbing plant type information from the headers and plant names. However, I'm only grabbing HYDRO references from the plant names because it often gets weird with certain plants having extremely misleading names like solar valley gas plant or something like that. In theory, I could create some customized code to address this, but I didn't yet because the `plant_type_2` column is already a bit precarious as noted above.

## What kind of review I'm looking for

Beyond any general feedback, here's what I'm looking for:

### Possible Errors
* How to address lingering uncertainty caused by programatic scraping and filling of the data (for FERC license, header plant type association, and mapping notes with no footnotes). Should I do a quick burn through all groups and manually check for outliers / create a list of manual overrides? Should I leave it as is? Should I add a caveat?
* Is there any other form of validation you would recommend adding to the code?
* Did you find any ways that the code miss-categorized or associated rows? I know there are some, and it would probably be a good idea to document as many as we can...

### Code Mechanics
* How efficient is the code? Could it be made faster?
* How well structured is the module? Do the functions make sense? Would you have done anything differently format-wise?
* Do you think I approached the problems correctly? I didn't use any machine learning because I dubbed it too much effort for the time being.
* Are there any little tricks that would make lines shorter / more legible? (I'm always trying to get better at this)

### Final Product
* How comfortable would you be using this data? Do you think we should exclude any components or include any disclaimers? What could engender the most confidence from users?
* How is the documentation? Can you follow it?

### Future Product
* Could any of this be moved to the transform module eventually/instead?

### CODE REVIEW

File: src/pudl/analysis/clean_up_ferc1.py [hunk 1]
```diff
@@ -0,0 +1,655 @@
...
+    'fuel_cell': ['fuel cell'],
+}
+
+
+########################################################################################
+# Helper Functions
+########################################################################################
+
+def expand_dict(dic):
+    """Change format of header_labels."""
```
Comment (bendnorman) [inline src/pudl/analysis/clean_up_ferc1.py:140]: How is this changing the format of `header_labels`?

File: src/pudl/analysis/clean_up_ferc1.py [hunk 2]
```diff
@@ -0,0 +1,655 @@
+"""Clean up the data from FERC Form 1.
+
+The data from FERC Form 1 is notoriously bad. The small generators table, for example,
+contains rows that represent different types of information: headers, values, and notes.
+It's like someone took a paper form, copied it into Microsoft Excel, and called it a
+day. Needless to say, it needs a lot of help before it can be used programatically for
+bulk analysis.
```
Comment (bendnorman) [inline src/pudl/analysis/clean_up_ferc1.py:7]: Is it possible to provide some examples of raw small generators records and the desired outputs after the transformations are applied?

File: src/pudl/analysis/clean_up_ferc1.py [hunk 3]
```diff
@@ -0,0 +1,655 @@
...
+# I have some FERC steam table flagging stuff I want to add but I'm going to do it
+# incrementally.
+
+########################################################################################
+# --------------------------------------------------------------------------------------
+#  S M A L L  G E N E R A T O R S  T A B L E
+# --------------------------------------------------------------------------------------
+########################################################################################
+# If these columns are nan, we can assume it is either a header row or isn't useful
+nan_cols = [
```
Comment (bendnorman) [inline src/pudl/analysis/clean_up_ferc1.py:34]: Constants are typically [all caps](https://www.python.org/dev/peps/pep-0008/#constants).

**Review by bendnorman — CHANGES_REQUESTED** [review]
This is looking good! I added a couple of style comments and requested some examples. 

Do we have a means of testing the outputs? There are many complicated transforms here and it could be helpful for comprehension and correctness to include some unit or validation tests.

File: src/pudl/analysis/clean_up_ferc1.py [hunk 4]
```diff
@@ -0,0 +1,655 @@
...
+def remove_bad_rows_sg(sg_df, show_removed=False):
+    """Test."""
+    # Remove utilities with all NAN rows because these won't contain anything meaningful
+    logger.info("Removing rows where an entire utility has reported NA in key columns")
+    sg_clean1 = (
+        sg_df.groupby('utility_id_ferc1')
+        .filter(lambda x: ~x[nan_cols].isna().all().all()))
+
+    # Remove rows with three or more dashes for plant name
+    logger.info("Removing rows with three or more dashes for plant name")
```
[the remark this answers is not in the document]
Comment (aesharpe) [reply src/pudl/analysis/clean_up_ferc1.py:187]: fair point!

File: src/pudl/analysis/clean_up_ferc1.py [hunk 5]
```diff
@@ -0,0 +1,655 @@
...
+    # Now deal with some outliers:
+
+    # Move total row values that are reported in notes row below to correct row above
+    # and null the values in the notes row.
+    num_cols = [
+        x for x in sg_df.select_dtypes(include=['float', 'Int64']).columns.tolist(
+        ) if x not in ['utility_id_ferc1', 'report_year', 'ferc_license_id']]
+    bad_row = sg_df[(sg_df['plant_name_ferc1'].str.contains(
+        "amounts are for")) & (sg_df['capacity_mw'] > 0)]
+    assert len(bad_row) == 1, 'found more bad rows than expected'
```
[the remark this answers is not in the document]
Comment (aesharpe) [reply src/pudl/analysis/clean_up_ferc1.py:220]: I think I fixed it / did a better job of explaining what is going on here.

File: src/pudl/analysis/clean_up_ferc1.py (hunk 1 above)
Comment (aesharpe) [reply src/pudl/analysis/clean_up_ferc1.py:140]: Did my best to explain in the doc string!

File: src/pudl/analysis/clean_up_ferc1.py (hunk 2 above)
Comment (aesharpe) [reply src/pudl/analysis/clean_up_ferc1.py:7]: done!

File: src/pudl/analysis/clean_up_ferc1.py (hunk 3 above)
Comment (aesharpe) [reply src/pudl/analysis/clean_up_ferc1.py:34]: good reminder

**Comment by TrentonBush** [comment]
It seems like Ben took on your "code mechanics" request, so I'll look for "possible errors".

My first comment is that I can't find the `clean_ferc1_analysis.ipynb` you mentioned. 😜 I'm on the latest commit of the `clean-small-gen-headers` branch. If you just forgot to commit it, could you use the `--no-verify` flag on `git commit`? That way pre-commit hooks won't strip out the cell outputs, and I can check if I get the same ones. I'm looking forward to seeing what you came up with!

**Comment by aesharpe** [comment]
>

**Comment by aesharpe** [comment]
> My first comment is that I can't find the `clean_ferc1_analysis.ipynb` you mentioned.

HAH, my bad... I added it in just now! @bendnorman feel free to take a look at this if you want to add more feedback. This will help a lot.

**Comment by TrentonBush** [comment]
### General comments responding to your requests:

#### Building Trust With Users: Validation
In the notebook, you compare performance vs Zane's manual labeling, which is great. I'd pull some of those comparison metrics straight into the documentation so people can see them easier. Personally I'd use sklearn.metrics to generate them rather than doing it manually, for the simple reason that sklearn is well tested whereas I might make a mistake. D.RY. and all.
Everyone knows our efforts are not going to be 100% accurate, but it is helpful to give guidance on approximately how accurate the estimates are: 60%? 99.9%? Giving an estimate does two things:
1. Demonstrates to our users that we have thought about these things, which builds confidence and trust
2. Gives users guidance on whether our data meets their specific accuracy needs

> Do you think I approached the problems correctly? I didn't use any machine learning because I dubbed it too much effort for the time being.
I think your rule-based algorithm was a good approach vs ML. It seems to have delivered good accuracy and will be more comprehensible to users.
One aspect I might have done differently is to build a kind of validation feedback loop earlier in the process (like after realizing the problem was too complex to solve in one day). You'll have to validate it at the end anyway, so building the validation process earlier just helps you during development. For example, make a function that compares the manual labels vs predicted. Then each time you tweak a regex rule or whatever, you re-run validation and see how the scores changed. It also helps to catch errors, which (at least for me) are common with regex in particular.

> Could any of this be moved to the transform module eventually/instead?
I think it should certainly be used in the transform process; the small plants table is pretty useless without your cleanup! If that's where we use it, it makes sense to move it there. Maybe in the next release?

**Comment by aesharpe** [comment]
> In the notebook, you compare performance vs Zane's manual labeling, which is great. I'd pull some of those comparison metrics straight into the documentation so people can see them easier.

I'll see about adding more validation, but ultimately the idea is that my programatic approach will replace Zane's manual approach.

**Comment by bendnorman** [comment]
What is the status of this PR?

**Comment by jdangerx** [comment]
@aesharpe is this PR still active?

**Comment by aesharpe** [comment]
Ah hey, thanks @bendnorman and @jdangerx for flagging this. I used this to help with the transforms for the small generators table within the new FERC transform framework, but now this is obsolete.

**Comment by jdangerx** [comment]
Sweet, gonna close this as a clean-up measure.
